src/shogun/classifier/Perceptron.h

```diff
@@ -71,29 +59,20 @@ class CPerceptron : public CLinearMachine
 		virtual const char* get_name() const { return "Perceptron"; }
 
 	protected:
-		/** registers and initializes parameters */
-			void init();
-
-		/** train classifier
-		 *
-		 * @param data training data (parameter can be avoided if distance or
-		 * kernel-based classifiers are used and distance/kernels are
-		 * initialized with train data)
-		 *
-		 * @return whether training was successful
-		 */
-		virtual bool train_machine(CFeatures* data=NULL);
+		virtual void init_model(CFeatures* data);
+		virtual void iteration();
 
 	protected:
 		/** learning rate */
 		float64_t learn_rate;
-		/** maximum number of iterations */
-		int32_t max_iter;
 
 	private:
+		/** Store output during training */
+		SGVector<float64_t> m_output;
```

Comment:
as an example for the end_training method, you could overload it here and then free the vector...i.e. overwrite it with empty vector